test/e2e/reconciler/clusterworkspacedeletion/controller_test.go

```diff
@@ -0,0 +1,197 @@
+				}
+
+				// create a configmap with finalizer
+				configmap := &corev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "test",
+						Namespace:  metav1.NamespaceDefault,
+						Finalizers: []string{"tenancy.kcp.dev/test-finalizer"},
+					},
+					Data: map[string]string{
+						"foo": "bar",
+					},
+				}
+
+				configmap, err = kubeClient.CoreV1().ConfigMaps(metav1.NamespaceDefault).Create(ctx, configmap, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create configmap in workspace %s", workspace.Name)
+
+				err = server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Delete(ctx, workspace.Name, metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete workspace %s", workspace.Name)
+
+				// The workspace condition should be updated since there is resource in the workspace pending
+				// finalization.
+				server.workspaceExpector.ExpectBefore(ctx, func(ctx context.Context) (bool, error) {
+					workspace, err := server.workspaceInformer.Lister().Get(key)
+					if err != nil {
+						return !apierrors.IsNotFound(err), err
+					}
+
+					if !conditions.IsTrue(workspace, tenancyv1alpha1.WorkspaceDeletionContentSuccess) {
+						return false, fmt.Errorf("expect true status for condition %s", tenancyv1alpha1.WorkspaceDeletionContentSuccess)
+					}
+
+					if !conditions.IsFalse(workspace, tenancyv1alpha1.WorkspaceContentClean) {
+						return false, fmt.Errorf("expect true status for condition %s", tenancyv1alpha1.WorkspaceContentClean)
+					}
+
+					return true, nil
+				}, 30*time.Second)
+
+				// clean finalizer to remove the configmap
+				configmap, err = kubeClient.CoreV1().ConfigMaps(metav1.NamespaceDefault).Get(ctx, configmap.Name, metav1.GetOptions{})
+				require.NoError(t, err, "failed to get configmap in workspace %s", workspace.Name)
+				configmap.Finalizers = []string{}
+				_, err = kubeClient.CoreV1().ConfigMaps(metav1.NamespaceDefault).Update(ctx, configmap, metav1.UpdateOptions{})
+				require.NoError(t, err, "failed to update configmap in workspace %s", workspace.Name)
+
+				// ensure workspace is removed
+				server.workspaceExpector.ExpectBefore(ctx, func(ctx context.Context) (bool, error) {
+					workspace, err := server.workspaceInformer.Lister().Get(key)
+					if apierrors.IsNotFound(err) {
+						return true, nil
+					}
+
+					return false, fmt.Errorf("workspace %s still exists", workspace.Name)
+				}, 60*time.Second)
+
+				// finally check if all resources has been removed.
+				nslist, err := kubeClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+				require.NoError(t, err, "failed to list namespaces in workspace %s", workspace.Name)
+				require.Equal(t, 0, len(nslist.Items))
+
+				cmlist, err := kubeClient.CoreV1().ConfigMaps(metav1.NamespaceAll).List(ctx, metav1.ListOptions{})
+				require.NoError(t, err, "failed to list configmaps in workspace %s", workspace.Name)
+				require.Equal(t, 0, len(cmlist.Items))
+			},
```

Comment:
would like to see another cluster-scoped resource in the test.

Reply:
Also add a second test case where an org workspace is deleted. That has the obvious complexity of nested deletions, but it also has virtual resources (`Workspace` which is mapped into the org workspace, but is actually served by the virtual workspace apiserver).

Reply:
cc @davidfestal – interesting topic when we map in virtual resources that are marked writeable in the fake CRD we put into the workspace, but readonly in the virtual workspace.

Reply:
also create a ns in the workspace, ensuring it is deleted and another test to delete org workspace.